### PR TITLE
[STRY/BE] REST API: Users & Passport Setup

### PR DESCRIPTION
**Pull request by ryankeegan**
Add passport, authenticate middleware, and CRUD for users.

### CODE REVIEW

File: source/routes/update/user.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
+            // Put newly updated document into the databse
```
Comment (b2wu) [inline source/routes/update/user.js:31]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              // Put newly updated document into the databse
```

File: source/routes/update/user.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
```
Comment (b2wu) [inline source/routes/update/user.js:30]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              response['password'] = hash;
```

File: source/routes/update/user.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
+                  res.send('success');
+                })
```
Comment (b2wu) [inline source/routes/update/user.js:35]: Codacy has a fix for the issue: [Expected indentation of 18 spaces but found 16.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                  })
```

File: source/routes/update/user.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      if (!req.body.updateField) {
+        throw new Error('MISSING UPDATE DATA');
+      }
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
```
Comment (b2wu) [inline source/routes/update/user.js:27]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              }
```

File: source/routes/update/user.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+module.exports = {
+  '/update/user': {
+    methods: ['post'],
+    middleware: [authenticate],
+    fn: function(req, res, next) {
+      if (!req.body.updateField) {
+        throw new Error('MISSING UPDATE DATA');
+      }
+      db.get(req.body.username)
+        .then((response) => {
```
Comment (b2wu) [inline source/routes/update/user.js:22]: Codacy has a fix for the issue: [Expected indentation of 10 spaces but found 8.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
          .then((response) => {
```

File: source/routes/update/user.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
```
Comment (b2wu) [inline source/routes/update/user.js:24]: Codacy has a fix for the issue: [Expected space(s) after "if".](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
            if (err) {
```

File: source/routes/update/user.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
+                  res.send('success');
+                })
+                .catch((err) => {
```
Comment (b2wu) [inline source/routes/update/user.js:36]: Codacy has a fix for the issue: [Expected indentation of 18 spaces but found 16.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                  .catch((err) => {
```

File: source/routes/update/user.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
+            // Put newly updated document into the databse
+            db.put(response)
```
Comment (b2wu) [inline source/routes/update/user.js:32]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              db.put(response)
```

File: source/routes/read/user.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// }
+// TODO: Remove? All we're returning is the username which is also the index queried to get the info
+module.exports = {
+  '/read/user': {
+    methods: ['get'],
+    middleware: [authenticate],
+    fn: function(req, res, next) {
+      db.get(req.body.username)
+          .then((response) => {
+            response.username = response._id;  // Response body should use 'username' not '_id'
```
Comment (b2wu) [inline source/routes/read/user.js:19]: Codacy has a fix for the issue: [Multiple spaces found before '// Response bo...'.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
            response.username = response._id; // Response body should use 'username' not '_id'
```

File: source/routes/update/user.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      }
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
```
Comment (b2wu) [inline source/routes/update/user.js:29]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              response['_id'] = req.body.username;
```

File: source/routes/update/user.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            response['password'] = hash;
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
+                  res.send('success');
+                })
+                .catch((err) => {
+                  console.log(err);
+                  res.send('error');
+                });
```
Comment (b2wu) [inline source/routes/update/user.js:39]: Codacy has a fix for the issue: [Expected indentation of 18 spaces but found 16.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                  });
```

File: source/routes/update/user.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
+                  res.send('success');
```
Comment (b2wu) [inline source/routes/update/user.js:34]: Codacy has a fix for the issue: [Expected indentation of 20 spaces but found 18.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                    res.send('success');
```

File: source/routes/update/user.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        throw new Error('MISSING UPDATE DATA');
+      }
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
+              res.send('error');
+            }
+            // Replace fields of the response document
```
Comment (b2wu) [inline source/routes/update/user.js:28]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              // Replace fields of the response document
```

File: source/routes/update/user.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    middleware: [authenticate],
+    fn: function(req, res, next) {
+      if (!req.body.updateField) {
+        throw new Error('MISSING UPDATE DATA');
+      }
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
+            if(err) {
+              console.log(err);
```
Comment (b2wu) [inline source/routes/update/user.js:25]: Codacy has a fix for the issue: [Expected indentation of 16 spaces but found 14.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                console.log(err);
```

File: source/routes/update/user.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+              res.send('error');
+            }
+            // Replace fields of the response document
+            response['_id'] = req.body.username;
+            response['password'] = hash;
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
```
Comment (b2wu) [inline source/routes/update/user.js:33]: Codacy has a fix for the issue: [Expected indentation of 18 spaces but found 16.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                  .then(() => {
```

File: source/routes/update/user.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  '/update/user': {
+    methods: ['post'],
+    middleware: [authenticate],
+    fn: function(req, res, next) {
+      if (!req.body.updateField) {
+        throw new Error('MISSING UPDATE DATA');
+      }
+      db.get(req.body.username)
+        .then((response) => {
+          bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
```
Comment (b2wu) [inline source/routes/update/user.js:23]: Codacy has a fix for the issue: [Expected indentation of 12 spaces but found 10.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
            bcrypt.hash(req.body.updateField.password, _saltRounds, function(err, hash) {
```

File: source/routes/update/user.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+            if(err) {
+              console.log(err);
+              res.send('error');
```
Comment (b2wu) [inline source/routes/update/user.js:26]: Codacy has a fix for the issue: [Expected indentation of 16 spaces but found 14.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
                res.send('error');
```

File: source/routes/update/user.js (hunk 6 above)
Comment (b2wu) [inline source/routes/update/user.js:24]: Codacy has a fix for the issue: [Expected indentation of 14 spaces but found 12.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7405452)
```suggestion
              if(err) {
```

File: source/routes/create/user.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+      requiredFields.forEach((jsonField, index) => {
+        if (!req.body[jsonField]) {
+          throw new Error('MISSING FIELD');
+        }
+      });
+
+      bcrypt.hash(req.body.password, _saltRounds, function(err, hash) {
+        db.post({
+          // Stores the username
+          // TODO: ensure username is unqiue
```
Comment (dino-inc) [inline source/routes/create/user.js:23]: The TODO should probably be finished or removed before merging.

File: source/routes/delete/session.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  '/delete/session': {
+    methods: ['delete'], // TODO: We may want to use post instead?
```
Comment (dino-inc) [inline source/routes/delete/session.js:3]: Another TODO... maybe say "for a future story" rather than TODO? It ultimately means the same thing, I suppose.

**Review by dino-inc — APPROVED** [review]
Besides the occasional TODOs, it seems perfectly functional and nothing particularly weird was done, as far as I can tell.

File: source/routes/create/user.js (hunk 18 above)
Comment (ryankeegan) [reply source/routes/create/user.js:23]: This TODO is actually done. I'll remove it

File: source/routes/delete/session.js (hunk 19 above)
Comment (ryankeegan) [reply source/routes/delete/session.js:3]: TODO is easier to find. Many text editors/IDEs provide features around 'TODO'. Using any other prefix would needlessly complicate things.

**Pull request merged**